### PR TITLE
Abstract surveys

### PR DESCRIPTION
**Pull request by staadecker**
This PR abstracts the concept of a survey to simplify the process of adding more surveys.

# Details

## Functionality changes

- [x] Moves `schema.version` to `metadata`.
- [x] Drop `schema.form` field

## Refactors

- [x] Gives each survey a key that is shared between frontend and backend in `util-shared-constants`
- [x] Refactors the redux store, submission route and survey page url to use the key
- [x] Refactors the backend to allow gravedigger and hospital surveys to use an api factory
- [x] Automatically generate frontend routes and home page buttons based on config
- [x] cleanup backend tests to avoid duplication

## Documentation 

- [x] Documents the process of adding a new survey

# Todo

- [x] Discuss the best way to manage schema and versions
- [x] Discuss the best way to handle the `schema.form` field.
- [x] Increment form versions (waiting on discussion for schema version)

# Testing (as of commit 59a3ef1)

- [x] All forms render properly
- [x] All forms submit properly
- [x] Upon submission, data is properly stored in the database
- [x] Form version is incremented

### CODE REVIEW

File: docs/Adding a survey.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+You need to add translations for the title of the survey & the FormIO content of the survey.
+
+The title translations need to be add in `src/translations/*/Surveys.json` with the same key you specified in step 2b (`i18nTitleKey`).
+
+The FormIO translations need to be added in new files in `src/translations/*/formio/`.
+You'll then have to import those files in `src/i18n.js` and add them to the `translation` parameter.
+Simply use the other surveys as a reference for this step.
+Note that FormIO shares translations for all forms so pick key names that don't conflict with other forms.
+
+## Step 3: Add to `backend`
```
Comment (Philip-Nicolas) [inline docs/Adding a survey.md:104]: vide

File: docs/Adding a survey.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+Testing for permissions is done globally for all surveys and doesn't need to be included.
+
+### Step 4: Deployment
+
+Test your changes locally, make sure the survey renders and can be submitted.
+Test your changes on staging using `yarn deploy-both`. Make sure the survey renders and can be submitted.
+Make sure the data is properly being stored in the database.
+Make a pull request and wait for your changes to be merged to master.
+
+### Step 5: Celebrate!
```
Comment (Philip-Nicolas) [inline docs/Adding a survey.md:217]: hilarant

File: packages/backend/src/surveys/initialHousehold/api.js [hunk 3]
```diff
@@ -62,4 +61,25 @@ async function initialSubmission(
   });
 }
 
-module.exports = { initialSubmission };
+const errorHandler = (apiFunction) => async (...args) => {
```
Comment (Philip-Nicolas) [inline packages/backend/src/surveys/initialHousehold/api.js:64]: document

File: packages/backend/test/surveys/permissions.test.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    ]);
+
+    await agent.post(endpoint).expect(400); // fail for non permission reasons (missing body)
+  });
+
+  // eslint-disable-next-line jest/expect-expect
+  it("should fail for a user without submit permissions", async () => {
+    const { agent } = await login(
+      app,
+      getAllPermissionsExcept(Permissions.submitForms)
```
Comment (Philip-Nicolas) [inline packages/backend/test/surveys/permissions.test.js:37]: does this method have a test?
also, is this the only usage?

Comment (staadecker) [reply packages/backend/test/surveys/permissions.test.js:37]: used multiple times throughout the tests and just added a unit test for the function

**Comment by Philip-Nicolas** [comment]
fully tested initial household survey

**Review by Philip-Nicolas — APPROVED** [review]

**Pull request merged**